### ISSUE TITLE
Message Handling Architecture Overhaul

### ISSUE THREAD

**Issue by sphaerophoria** [qTox/qTox#5691]
Hi all. I'd like to propose an overhaul to the message handling path in qTox. Let me know what you think :)

# Motivation

As some of you may know I've been working on adding extension support to qTox. I wanted a super clean proof of concept implementation so people could see how simple it was to integrate with a client. Unfortunately the qTox message handling path is not as clean as I would have hoped. Figuring out where to hook in the extension manager is a bit of a pain. Some problems with the current architecture I noticed

* Friend chat messages are split in ChatForm, but group messages are split in core
* History messages are added on reception by Widget, but outgoing messages are added by ChatForm
* ChatForm/GroupChatForm both do the same thing with messages... we can consolidate some duplicated logic here
* While we're on the topic of ChatForm/GroupChatForm, GroupChatForm emits a signal that it wants to send a message, but ChatForm just calls down to Core itself resulting in two completely different paths for a very similar concept
* Why is chat history owned by a gui chat form?
* We have no model level abstraction of messages, this results in us looking back through our chatlog and looking at gui rendered messages in order to do logic
* Currently the only places chat messages are stored are in history and in the chatform. It would be nice to have some mid level abstraction so we can hold more messages in ram without wasting tons of ram rendering them all at once
* Timestamps generated in several places

My proposal is to split as much out of Widget and (Generic/Group)ChatForm as possible and do logic on a mid level representation. Once this is done implementing the work I have planned around extensions will be significantly easier

# Current design

![image](https://user-images.githubusercontent.com/1069811/58477177-c9171c80-8107-11e9-9dd5-03fc98be071d.png)

[Source](http://www.nomnoml.com.s3-website-eu-west-1.amazonaws.com/#view/%5BWidget%7CMessage%20related%20responsibilities%3A%0A-%20Notify%20on%20friend%20message%0A-%20Notify%20on%20group%20message%0A-%20Manage%20chat%20forms%0A-%20Generate%20timestamps%20for%20history%0A-%20Add%20received%20messages%20to%20history%0A%5D%0A%0A%5BChatForm%7CMesage%20related%20responsibilities%3A%0A-%20Manage%20offline%20messages%0A-%20Load%20history%0A-%20Send%20unsent%20messages%20when%20loaded%20from%20history%0A-%20Split%20messages%20before%20sending%0A-%20Render%20messages%20to%20the%20chat%20log%0A-%20Parse%20chat%20entry%20for%20action%20messages%0A-%20Generate%20timestamps%20for%20chat%20log%20(note%20that%20these%20are%0A%20%20different%20than%20the%20timestamps%20generated%20for%20history)%0A-%20Detect%20if%20any%20received%20messages%20are%20for%20this%20instance%20of%0A%20%20ChatForm%0A%5D%0A%0A%5BGroupChatForm%7CMesasge%20related%20responsibilities%3A%0A-%20Render%20messages%20to%20the%20chat%20log%0A-%20Parse%20chat%20entry%20for%20action%20messages%0A%5D%0A%0A%5BGenericChatForm%7CMesage%20related%20responsibilities%3A%0A-%20Handles%20actual%20creation%20of%20messages%20%0A%20%20GroupChatForm%20and%20ChatForm%20want%20to%0A%20%20render%2C%20but%20not%20the%20logic%20of%20when%20to%20render%20them%0A%5D%0A%0A%5BCore%7CMesasge%20related%20responsibilities%3A%0A-%20Split%20group%20messages%20before%20sending%0A-%20Send%2FReceive%20group%2Ffriend%20messages%0A%20%20from%20toxcore%0A%5D%0A%0A%5BGenericChatForm%5D%20%3C%3A--%20%5BChatForm%5D%0A%5BGenericChatForm%5D%20%3C%3A--%20%5BGroupChatForm%5D%0A%0A%5BWidget%5D%20%2B--%200..%5BChatForm%5D%0A%5BWidget%5D%20%2B--%200..%5BGroupChatForm%5D%0A%0A%5BCore%5D%20--o%20%5BChatForm%5D%0A%5BCore%5D%20--o%20%5BWidget%5D%0A%0A%5BHistory%5D%20--o%20%5BWidget%5D%0A%5BHistory%5D%20--o%20%5BChatForm%5D%0A%5BHistory%5D%20--%20%5B%3Cnote%3E%20Written%20to%20by%20Widget%20on%20messageReceived%0A%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20and%20by%20ChatForm%20on%20message%20sent.%20Why%20not%0A%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20at%20least%20have%20chatform%20do%20both%3F%5D%0A%0A%0A%0A)

# Core Design

The core of the new design resolves around the concept of a Mesasge and an IMessageDispatcher. Currently Widget seems to own most of our classes in this area, so for now we'll give Widget a message dispatcher per conversation.

The message dispatchers would read messages from both Core and our ChatForms and decide where they should go. Any class that wants to do an action on messages (either sent or received) just hook up slots to the dispatcher.

For example notifiers will hook up to the "messageReceived" signal, but chatForm will hook up to the "messageSent, messageReceived, and messageCompleted" signals.

Overall design is as follows

![image](https://user-images.githubusercontent.com/1069811/58477202-d9c79280-8107-11e9-8637-edf6ac6ad0d1.png)

[Source](http://www.nomnoml.com.s3-website-eu-west-1.amazonaws.com/#view/%0A%5BMessage%7C%0A%2BisAction%20%3A%20bool%0A%2Bcontent%20%3A%20QString%0A%2Btimestamp%20%3A%20QDateTime%0A%2Bmetadata%20%3A%20std%3A%3Avector%3CMetadata%3E%5D%0A%0A%5BMessageProcessor%7C%0AResponsibilities%3A%0A-%20Split%20messages%20for%20Core%0A-%20Add%20timestamp%20to%20messages%0A-%20Detect%20mentions%5D%0A%0A%5BIMessageDispatcher%7C%0AResponsibilities%3A%0A-%20Send%20messages%20to%20core%0A-%20Emit%20signals%20when%20messages%20received%5D%0A%0A%5BWidget%7C%0AResponsibilities%3A%0A-%20Hold%20messaging%20classes%5D%0A%0A%5BFriendMessageDispatcher%7C%0AResponsibilities%3A%0A-%20Process%20friend%20messages%20from%20core%0A-%20Implement%20IMessageDispatcher%5D%0A%0A%5BGroupMessageDispatcher%7C%0AResponsibilities%3A%0A-%20Process%20friend%20messages%20from%20core%0A-%20Implement%20IMessageDispatcher%5D%0A%0A%5BNotificationManager%7C%0AResponsibilities%3A%0A-%20Manage%20split%20between%20snorify%20messages%20%0A%20%20and%20Qt%20style%20alerts%0A%5D%0A%0A%5BGroupNotifier%7C%0AResponsibilities%3A%0A-%20Process%20Messages%20given%20by%20GroupMessageDispatcher%0A%20%20and%20send%20them%20to%20the%20NotificationManager%5D%0A%20%20%0A%5BFriendNotifier%7C%0AResponsibilities%3A%0A-%20Process%20Messages%20given%20by%20FriendMessageDispatcher%0A%20%20and%20send%20them%20to%20the%20NotificationManager%5D%0A%20%20%0A%5BChatHistory%7C%0AResponsibilities%3A%0A-%20Process%20Messages%20given%20by%20FriendMessageDispatcher%0A%20%20and%20send%20them%20to%20history%0A%5D%0A%0A%5BGenericChatForm%7C%0AResponsibilities%3A%0A-%20Render%20messages%20from%20Message%20dispatchers%0A-%20Send%20messages%20to%20the%20dispatchers%5D%0A%0A%5BWidget%5D%20%2B-%3E%200..%5BFriendMessageDispatcher%5D%0A%5BWidget%5D%20%2B-%3E%200..%5BGroupMessageDispatcher%5D%0A%5BWidget%5D%20%2B-%3E%201%5BNotificationManager%5D%0A%5BWidget%5D%20%2B-%3E%200..%5BGenericChatForm%5D%0A%5BWidget%5D%20%2B-%3E%200..%5BFriendNotifier%5D%0A%5BWidget%5D%20%2B-%3E%200..%5BGroupNotifier%5D%0A%5BWidget%5D%20%2B-%3E%200..%5BChatHistory%5D%0A%0A%5BIMessageDispatcher%5D%20%3C%3A--%20%5BFriendMessageDispatcher%5D%0A%5BIMessageDispatcher%5D%20%3C%3A--%20%5BGroupMessageDispatcher%5D%0A%0A%5BFriendMessageDispatcher%5D%20o-%3E%20%5BMessageProcessor%5D%0A%5BGroupMessageDispatcher%5D%20o-%3E%20%5BMessageProcessor%5D%0A%0A%5BGroupNotifier%5D%20o-%3E%20%5BIMessageDispatcher%5D%0A%5BGroupNotifier%5D%20o-%3E%20%5BNotificationManager%5D%0A%0A%5BFriendNotifier%5D%20o-%3E%20%5BIMessageDispatcher%5D%0A%5BFriendNotifier%5D%20o-%3E%20%5BNotificationManager%5D%0A%0A%5BChatHistory%5D%20o-%3E%20%5BHistory%5D%0A%5BChatHistory%5D%20o-%3E%20%5BIMessageDispatcher%5D%0A%0A%5BGenericChatForm%5D%20o-%3E%20%5BIMessageDispatcher%5D%0A%5BGenericChatForm%5D%20o-%3E%20%5BChatHistory%5D%0A%0A%5BMessageProcessor%5D%20--%20%5BMessage%5D%0A%0A%23direction%3A%20right)

Benefits of above design
* Separation of responsibilities. Each class is short and sweet
* Significantly more easily testable. We can mock out the message dispatcher and see how any of the associated classes relate to incoming/outcoing messages
* Extending for new features becomes a lot simpler. We can work with "Messages" instead of parsing ChatMessage::Ptrs as we were previously doing (see old implementation of OfflineMsgEngine)

**Comment by sudden6** [comment]
@Diadlo ping, I think this might be in line with what you are working on

@sphaerophoria You mention a `NotificationManager`, so you might also take a look at this project: https://github.com/qTox/qTox/projects/4

In general I like the idea of making qTox messages more flexible, but I didn't have the time to take a deeper look at your proposal so far.

**Comment by sphaerophoria** [comment]
@sudden6 that notification center project is very close to what I had in mind. Are you planning on working that soon? If so I can do a rudimentary hookup for that section of code and just handle the message dispatchers for now (which are mostly done, notifications were one section I wasn't happy with the cleanliness of what I had so far).

Really the only reason I was going to bring that into scope was because I didn't like how I had hooked up the current notifications to widget, but I don't really need to touch those if a cleanup is coming in the near future anyways :).
